### PR TITLE
Add field based locking for objects and records

### PR DESCRIPTION
**Pull request by vinok88**
## Approach

```
type Person object {
    string stars;

    public function update(string s) {
            lock {
                   stars = stars + s;
              }     
 }
};
```

- Access to the stars variable within the update function from the same object instance will be sequential.
- Parallel access to the stars variable in the update function from different object instances will be allowed.

```
function incrementParam(Student param, Person person) {
   lock {
        param.score = param.score + i;
        person.address.street  = "palmgrove"
    }
}
```
- Parallel access to the score field by passing the same Student instance will not be allowed.
- When there is nested record access, both fields will be locked. In the above example, both address field of person and the street field of address record will be locked.

## Known issues
```
function incrementParam(Student param) {
   lock {
        Student inLockObj = param;
        inLockObj.score = 10;
        inLockObj = new;
        inLockObj.age = 78;
    }
}
```

-  will not work here based on the inLockObj.score field access. The reason is, here the direct field access is done to a local variable. We need to do a data flow analysis in order to identify this is actually referring to an outer parameter.
- locking will not work if the reference of a ref type is changed within the lock block. For example, if you reinitiate (new) a passed in record within the lock, the locking will be done only for the ref that was available at the entry point of the lock block. In the example above, inLockObj.age will not be locked since it was not available (the reference) while entering the lock block.

### CODE REVIEW

File: bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/CPU.java [hunk 1]
```diff
@@ -2790,13 +2796,42 @@ private static boolean handleVariableLock(WorkerExecutionContext ctx, BType[] ty
...
+
+        }
+
+        if (varRegs.length > varCount) {
+            if (ctx.localProps == null) {
+                ctx.localProps = new HashMap<>();
+            }
+
+            ctx.localProps.putIfAbsent(Constants.LOCK_VARS, new HashMap<UUID, Stack<VarLock>>());
+            HashMap locks = (HashMap) ctx.localProps.get(Constants.LOCK_VARS);
```
Comment (lankavitharana) [inline bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/CPU.java]: shall we use just Map here?

File: bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/CPU.java [hunk 2]
```diff
@@ -2790,13 +2796,42 @@ private static boolean handleVariableLock(WorkerExecutionContext ctx, BType[] ty
...
             }
+
+        }
+
+        if (varRegs.length > varCount) {
+            if (ctx.localProps == null) {
+                ctx.localProps = new HashMap<>();
+            }
+
+            ctx.localProps.putIfAbsent(Constants.LOCK_VARS, new HashMap<UUID, Stack<VarLock>>());
```
Comment (lankavitharana) [inline bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/CPU.java]: we can simply put i guess

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/codegen/CodeGenerator.java [hunk 3]
```diff
@@ -2840,13 +2842,18 @@ public void visit(BLangWhile whileNode) {
...
 
     public void visit(BLangLock lockNode) {
-        if (lockNode.lockVariables.isEmpty()) {
+        if (lockNode.lockVariables.isEmpty() && lockNode.fieldVariables.isEmpty()) {
             this.genNode(lockNode.body, this.env);
             return;
         }
+
+        //remove objects initialized within lock
+        lockNode.fieldVariables.keySet().removeIf(symbol -> symbol.varIndex == null);
```
Comment (lankavitharana) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/codegen/CodeGenerator.java]: shall we check this a bit? not sure checking null is ok

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/codegen/CodeGenerator.java [hunk 4]
```diff
@@ -2855,21 +2862,88 @@ public void visit(BLangLock lockNode) {
...
+            Set<BLangStructFieldAccessExpr> expressions = entry.getValue();
+
+            int pkgRefCPIndex = addPackageRefCPEntry(currentPkgInfo, symbol.pkgID);
+
+            for (BLangStructFieldAccessExpr expr : expressions) {
+                operands[i++] = getOperand(pkgRefCPIndex);
+                operands[i++] = expr.expr.regIndex;
+
+                int fieldNameCPEntry = addUTF8CPEntry(currentPkgInfo, (String) ((BLangLiteral) expr.indexExpr).value);
+                operands[i++] = getOperand(fieldNameCPEntry);
```
Comment (lankavitharana) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/codegen/CodeGenerator.java]: we can directly use index here

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/codegen/CodeGenerator.java (hunk 3 above)
Comment (vinok88) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/codegen/CodeGenerator.java]: I have checked by debugging and it does not cause a removal of any unintended variables other than the variables defined within a lock. But, I agree we should have a better way to identify this.

**Pull request merged**